core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java

```diff
@@ -162,119 +148,138 @@ public int getTopologyId() {
-            Object key = ice.getKey();  //todo [anistor] should we check for expired entries?
-            int segmentId = keyPartitioner.getSegment(key);
-            if (segments.contains(segmentId) && !ice.isL1Entry()) {
-               InternalCacheEntry entry = mapEntryFromDataContainer.apply(ice, entryFactory);
-               if (entry != null) {
-                  sendEntry(entry, segmentId);
-               }
-            }
+         if (subscription != null) {
+            throw new IllegalStateException("execute() may only be called once once");
          }
-
-         AdvancedCacheLoader<Object, Object> stProvider = persistenceManager.getStateTransferProvider();
-         if (stProvider != null) {
-            try {
-               Flowable.fromPublisher(stProvider.entryPublisher(k -> !dataContainer.containsKey(k), true, true))
-                     .blockingForEach(me -> {
-                        int segmentId = keyPartitioner.getSegment(me.getKey());
-                        if (segments.contains(segmentId)) {
-                           try {
-                              InternalCacheEntry entry = mapEntryFromStore.apply(me, entryFactory);
-                              if (entry != null) {
-                                 sendEntry(entry, segmentId);
-                              }
-                           } catch (CacheException e) {
-                              log.failedLoadingValueFromCacheStore(me.getKey(), e);
-                           }
-                        }
-                     });
-            } catch (CacheException e) {
-               log.failedLoadingKeysFromCacheStore(e);
-            }
+         AtomicReference<List<InternalCacheEntry<Object, Object>>> batchRef =
+            new AtomicReference<>(Collections.emptyList());
+         subscription =
+            Flowable.concat(publishDataContainerEntries(), publishStoreEntries())
+                    .buffer(chunkSize)
+                    .concatMapCompletable(
+                       // Send the previous batch, not the current one
+                       // This allows us to mark all the segments as finished in the same RPC with the last batch
+                       entries -> {
+                          List<InternalCacheEntry<Object, Object>> previousBatch = batchRef.getAndSet(entries);
+                          if (previousBatch.isEmpty())
+                             return Completable.complete();
+
+                          return completionStageToCompletable().apply(sendEntries(previousBatch, false));
+                       })
+                    .subscribe(() -> {
+                       // Send the remaining entries and mark all the segments as finished
+                       List<InternalCacheEntry<Object, Object>> previousBatch = batchRef.get();
+                       sendEntries(previousBatch, true)
+                          .whenComplete((ignored, throwable) -> {
+                             taskFuture.complete(null);
```

Comment:
To be safe we should probably invoke `completeExceptionally` here if `throwable` was not null.

Reply:
`sendEntries()` never throws, but ok

Reply:
I always like to think that as well, until it does :) And then we lost the exception, making it very difficult to figure out what is going on.